common/src/main/java/org/apache/celeborn/common/network/ssl/SSLFactory.java

```diff
@@ -0,0 +1,407 @@
+     * @param trustStoreReloadingEnabled Whether trust store reloading is enabled
+     * @param trustStoreReloadIntervalMs The interval at which to reload the trust store file
+     * @return The builder object
+     */
+    public Builder trustStore(
+        File trustStore,
+        String trustStorePassword,
+        boolean trustStoreReloadingEnabled,
+        int trustStoreReloadIntervalMs) {
+      this.trustStore = trustStore;
+      this.trustStorePassword = trustStorePassword;
+      this.trustStoreReloadingEnabled = trustStoreReloadingEnabled;
+      this.trustStoreReloadIntervalMs = trustStoreReloadIntervalMs;
+      return this;
+    }
+
+    /**
+     * Builds our {@link SSLFactory}
+     *
+     * @return The built {@link SSLFactory}
+     */
+    public SSLFactory build() {
+      return new SSLFactory(this);
+    }
+  }
+
+  /**
+   * Returns an initialized {@link SSLContext}
+   *
+   * @param requestedProtocol The requested protocol to use
+   * @param keyManagers The list of key managers to use
+   * @param trustManagers The list of trust managers to use
+   * @return The built {@link SSLContext}
+   * @throws GeneralSecurityException
+   */
+  private static SSLContext createSSLContext(
+      String requestedProtocol, KeyManager[] keyManagers, TrustManager[] trustManagers)
+      throws GeneralSecurityException {
+    SSLContext sslContext = SSLContext.getInstance(requestedProtocol);
+    sslContext.init(keyManagers, trustManagers, null);
+    return sslContext;
+  }
+
+  /**
+   * Creates a new {@link SSLEngine}. Note that currently client auth is not supported
+   *
+   * @param isClient Whether the engine is used in a client context
+   * @param allocator The {@link ByteBufAllocator to use}
+   * @return A valid {@link SSLEngine}.
+   */
+  public SSLEngine createSSLEngine(boolean isClient, ByteBufAllocator allocator) {
+    SSLEngine engine = createEngine(isClient, allocator);
+    engine.setUseClientMode(isClient);
+    engine.setWantClientAuth(true);
+    engine.setEnabledProtocols(enabledProtocols(engine, requestedProtocol));
+    engine.setEnabledCipherSuites(enabledCipherSuites(engine, requestedCiphers));
+    return engine;
+  }
+
+  private SSLEngine createEngine(boolean isClient, ByteBufAllocator allocator) {
+    return jdkSslContext.createSSLEngine();
+  }
+
+  private static final X509Certificate[] EMPTY_CERT_ARRAY = new X509Certificate[0];
+
+  private static TrustManager[] credulousTrustStoreManagers() {
+    return new TrustManager[] {
+      new X509TrustManager() {
+        @Override
+        public void checkClientTrusted(X509Certificate[] x509Certificates, String s)
+            throws CertificateException {}
+
+        @Override
+        public void checkServerTrusted(X509Certificate[] x509Certificates, String s)
+            throws CertificateException {}
+
+        @Override
+        public X509Certificate[] getAcceptedIssuers() {
+          return EMPTY_CERT_ARRAY;
```

Comment:
Note: This fixes a bug in spark, and I will backport it there.

Reply:
I assume this surfaced as part of the usage of `wantClientAuth`? What was the bug?

Reply:
The api expects an empty array and not null to be returned.
As part of testing various combinations, this surfaced, yes: though I do t recall if it was specifically related to want auth - it probably was

We need something similar to ssl connectivity test in Spark as well, which can help drive the matrix of combinations for testing  (there are a lot more combinations in Spark)